File: 2498-smallest-subarrays-with-maximum-bitwise-or/smallest-subarrays-with-maximum-bitwise-or.go
func smallestSubarrays(nums []int) []int {
	n := len(nums)
	lastSeen := make([]int, 30) // Track last seen index for each bit
	res := make([]int, n)

	// Initialize result with 1 since minimum subarray length is 1
	for i := range res {
		res[i] = 1
	}

	// Traverse from right to left
	for i := n - 1; i >= 0; i-- {
		for bit := 0; bit < 30; bit++ {
			if (nums[i] & (1 << bit)) > 0 {
				lastSeen[bit] = i
			}
			if lastSeen[bit] > 0 {
				res[i] = max(res[i], lastSeen[bit]-i+1)
			}
		}
	}
	return res
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: 2498-smallest-subarrays-with-maximum-bitwise-or/smallest-subarrays-with-maximum-bitwise-or.py
class Solution:
    def smallestSubarrays(self, nums: List[int]) -> List[int]:
        n = len(nums)
        last_seen = [0] * 30  # To store the last seen index for each bit
        res = [1] * n  # At minimum, the subarray length is 1

        for i in range(n - 1, -1, -1):  # Traverse from right to left
            for bit in range(30):  # For each bit from 0 to 29
                if nums[i] & (1 << bit):
                    last_seen[bit] = i  # Update last seen index of this bit
                res[i] = max(res[i], last_seen[bit] - i + 1)  # Update result
        return res
